File: apps/desktop/src/lib/api/focus.ts
import apiClient from './client';

// ============================================================================
// Types
// ============================================================================

export interface FocusSettings {
  id: string;
  defaultBlockedApps: string[];
  defaultBlockedWebsites: string[];
  allowedApps: string[];
  allowedWebsites: string[];
  autoStartFromCalendar: boolean;
  autoDetectGaps: boolean;
  minGapMinutes: number;
  focusDurationMinutes: number;
  breakDurationMinutes: number;
  longBreakDurationMinutes: number;
  sessionsBeforeLongBreak: number;
  breakRemindersEnabled: boolean;
  blockingMode: 'soft' | 'hard' | 'strict';
  workStartTime: string;
  workEndTime: string;
  workDays: number[];
  totalFocusMinutes: number;
  totalDistractionsBlocked: number;
  currentStreakDays: number;
  longestStreakDays: number;
}

export interface FocusSettingsUpdate {
  defaultBlockedApps?: string[];
  defaultBlockedWebsites?: string[];
  allowedApps?: string[];
  allowedWebsites?: string[];
  autoStartFromCalendar?: boolean;
  autoDetectGaps?: boolean;
  minGapMinutes?: number;
  focusDurationMinutes?: number;
  breakDurationMinutes?: number;
  longBreakDurationMinutes?: number;
  sessionsBeforeLongBreak?: number;
  breakRemindersEnabled?: boolean;
  blockingMode?: 'soft' | 'hard' | 'strict';
  bypassPassword?: string;
  workStartTime?: string;
  workEndTime?: string;
  workDays?: number[];
}

export interface FocusBlock {
  id: string;
  startTime: string;
  endTime: string;
  durationMinutes: number;
  title: string;
  status: 'scheduled' | 'active' | 'completed' | 'paused' | 'cancelled';
  blockingEnabled: boolean;
  blockedApps: string[];
  blockedWebsites: string[];
  completedMinutes: number;
  successRate: number | null;
  distractionsBlocked: number;
}

export interface CalendarGap {
  startTime: string;
  endTime: string;
  durationMinutes: number;
  qualityScore: number;
}

export interface FocusSuggestion {
  startTime: string;
  endTime: string;
  durationMinutes: number;
  reason: string;
  priority: 'high' | 'medium' | 'low';
}

export interface FocusStats {
  totalFocusMinutes: number;
  totalSessions: number;
  completedSessions: number;
  averageSessionMinutes: number;
  totalDistractionsBlocked: number;
  averageSuccessRate: number;
  currentStreak: number;
  longestStreak: number;
  dailyStats: Array<{
    date: string;
    focusMinutes: number;
    sessions: number;
    distractionsBlocked: number;
  }>;
}

export interface CreateFocusBlockRequest {
  startTime: string;
  endTime: string;
  title?: string;
  blockingEnabled?: boolean;
  syncToCalendar?: boolean;
}

// ============================================================================
// Focus Settings API
// ============================================================================

export async function getFocusSettings(): Promise<FocusSettings> {
  const response = await apiClient.get('/api/focus/settings');
  return transformFocusSettings(response.data);
}

export async function updateFocusSettings(settings: FocusSettingsUpdate): Promise<FocusSettings> {
  // Transform camelCase to snake_case for the backend
  const payload = {
    default_blocked_apps: settings.defaultBlockedApps,
    default_blocked_websites: settings.defaultBlockedWebsites,
    allowed_apps: settings.allowedApps,
    allowed_websites: settings.allowedWebsites,
    auto_start_from_calendar: settings.autoStartFromCalendar,
    auto_detect_gaps: settings.autoDetectGaps,
    min_gap_minutes: settings.minGapMinutes,
    focus_duration_minutes: settings.focusDurationMinutes,
    break_duration_minutes: settings.breakDurationMinutes,
    long_break_duration_minutes: settings.longBreakDurationMinutes,
    sessions_before_long_break: settings.sessionsBeforeLongBreak,
    break_reminders_enabled: settings.breakRemindersEnabled,
    blocking_mode: settings.blockingMode,
    bypass_password: settings.bypassPassword,
    work_start_time: settings.workStartTime,
    work_end_time: settings.workEndTime,
    work_days: settings.workDays,
  };

  // Remove undefined values
  const cleanPayload = Object.fromEntries(
    Object.entries(payload).filter(([, v]) => v !== undefined)
  );

  const response = await apiClient.put('/api/focus/settings', cleanPayload);
  return transformFocusSettings(response.data);
}

// ============================================================================
// Focus Blocks API
// ============================================================================

export async function getFocusBlocks(hoursAhead: number = 24): Promise<FocusBlock[]> {
  const response = await apiClient.get('/api/focus/blocks', {
    params: { hours_ahead: hoursAhead },
  });
  return response.data.map(transformFocusBlock);
}

export async function createFocusBlock(request: CreateFocusBlockRequest): Promise<FocusBlock> {
  const payload = {
    start_time: request.startTime,
    end_time: request.endTime,
    title: request.title || 'Focus Time',
    blocking_enabled: request.blockingEnabled ?? true,
    sync_to_calendar: request.syncToCalendar ?? false,
  };
  const response = await apiClient.post('/api/focus/blocks', payload);
  return transformFocusBlock(response.data);
}

export async function deleteFocusBlock(blockId: string): Promise<{ message: string }> {
  const response = await apiClient.delete(`/api/focus/blocks/${blockId}`);
  return response.data;
}

// ============================================================================
// Active Session API
// ============================================================================

export async function getActiveSession(): Promise<FocusBlock | null> {
  const response = await apiClient.get('/api/focus/active');
  return response.data ? transformFocusBlock(response.data) : null;
}

export async function startFocusSession(blockId?: string): Promise<FocusBlock> {
  const payload = blockId ? { block_id: blockId } : {};
  const response = await apiClient.post('/api/focus/start', payload);
  return transformFocusBlock(response.data);
}

export async function endFocusSession(blockId: string): Promise<FocusBlock> {
  const response = await apiClient.post(`/api/focus/end/${blockId}`);
  return transformFocusBlock(response.data);
}

export async function pauseFocusSession(blockId: string): Promise<FocusBlock> {
  const response = await apiClient.post(`/api/focus/pause/${blockId}`);
  return transformFocusBlock(response.data);
}

// ============================================================================
// Quick Actions
// ============================================================================

export async function quickStartFocus(durationMinutes: number = 50): Promise<FocusBlock> {
  const response = await apiClient.post('/api/focus/quick-start', null, {
    params: { duration_minutes: durationMinutes },
  });
  return transformFocusBlock(response.data);
}

export async function scheduleFromSuggestion(
  startTime: string,
  endTime: string
): Promise<FocusBlock> {
  const response = await apiClient.post('/api/focus/schedule-from-suggestion', null, {
    params: { start_time: startTime, end_time: endTime },
  });
  return transformFocusBlock(response.data);
}

// ============================================================================
// Suggestions & Gaps
// ============================================================================

export async function getCalendarGaps(
  daysAhead: number = 7,
  minGapMinutes: number = 30
): Promise<CalendarGap[]> {
  const response = await apiClient.get('/api/focus/gaps', {
    params: { days_ahead: daysAhead, min_gap_minutes: minGapMinutes },
  });
  return response.data.map((gap: Record<string, unknown>) => ({
    startTime: gap.start_time as string,
    endTime: gap.end_time as string,
    durationMinutes: gap.duration_minutes as number,
    qualityScore: gap.quality_score as number,
  }));
}

export async function getFocusSuggestions(daysAhead: number = 7): Promise<FocusSuggestion[]> {
  const response = await apiClient.get('/api/focus/suggestions', {
    params: { days_ahead: daysAhead },
  });
  return response.data.map((s: Record<string, unknown>) => ({
    startTime: s.start_time as string,
    endTime: s.end_time as string,
    durationMinutes: s.duration_minutes as number,
    reason: s.reason as string,
    priority: s.priority as string,
  }));
}

// ============================================================================
// Stats & Tracking
// ============================================================================

export async function getFocusStats(days: number = 7): Promise<FocusStats> {
  const response = await apiClient.get('/api/focus/stats', {
    params: { days },
  });
  const data = response.data;
  return {
    totalFocusMinutes: data.total_focus_minutes || 0,
    totalSessions: data.total_sessions || 0,
    completedSessions: data.completed_sessions || 0,
    averageSessionMinutes: data.average_session_minutes || 0,
    totalDistractionsBlocked: data.total_distractions_blocked || 0,
    averageSuccessRate: data.average_success_rate || 0,
    currentStreak: data.current_streak || 0,
    longestStreak: data.longest_streak || 0,
    dailyStats: (data.daily_stats || []).map((d: Record<string, unknown>) => ({
      date: d.date as string,
      focusMinutes: d.focus_minutes as number,
      sessions: d.sessions as number,
      distractionsBlocked: d.distractions_blocked as number,
    })),
  };
}

export async function recordDistractionBlocked(
  blockId: string,
  appOrSite: string
): Promise<{ message: string }> {
  const response = await apiClient.post('/api/focus/distraction-blocked', {
    block_id: blockId,
    app_or_site: appOrSite,
  });
  return response.data;
}

export async function checkAutoStart(): Promise<{
  shouldStart: boolean;
  event: Record<string, unknown> | null;
}> {
  const response = await apiClient.get('/api/focus/check-auto-start');
  return {
    shouldStart: response.data.should_start,
    event: response.data.event,
  };
}

// ============================================================================
// Transform Functions (snake_case to camelCase)
// ============================================================================

function transformFocusSettings(data: Record<string, unknown>): FocusSettings {
  return {
    id: data.id as string,
    defaultBlockedApps: data.default_blocked_apps as string[],
    defaultBlockedWebsites: data.default_blocked_websites as string[],
    allowedApps: data.allowed_apps as string[],
    allowedWebsites: data.allowed_websites as string[],
    autoStartFromCalendar: data.auto_start_from_calendar as boolean,
    autoDetectGaps: data.auto_detect_gaps as boolean,
    minGapMinutes: data.min_gap_minutes as number,
    focusDurationMinutes: data.focus_duration_minutes as number,
    breakDurationMinutes: data.break_duration_minutes as number,
    longBreakDurationMinutes: data.long_break_duration_minutes as number,
    sessionsBeforeLongBreak: data.sessions_before_long_break as number,
    breakRemindersEnabled: data.break_reminders_enabled as boolean,
    blockingMode: data.blocking_mode as 'soft' | 'hard' | 'strict',
    workStartTime: data.work_start_time as string,
    workEndTime: data.work_end_time as string,
    workDays: data.work_days as number[],
    totalFocusMinutes: data.total_focus_minutes as number,
    totalDistractionsBlocked: data.total_distractions_blocked as number,
    currentStreakDays: data.current_streak_days as number,
    longestStreakDays: data.longest_streak_days as number,
  };
}

function transformFocusBlock(data: Record<string, unknown>): FocusBlock {
  return {
    id: data.id as string,
    startTime: data.start_time as string,
    endTime: data.end_time as string,
    durationMinutes: data.duration_minutes as number,
    title: data.title as string,
    status: data.status as FocusBlock['status'],
    blockingEnabled: data.blocking_enabled as boolean,
    blockedApps: data.blocked_apps as string[],
    blockedWebsites: data.blocked_websites as string[],
    completedMinutes: data.completed_minutes as number,
    successRate: data.success_rate as number | null,
    distractionsBlocked: data.distractions_blocked as number,
  };
}

File: apps/backend/app/api/routes/focus.py
"""
Focus Mode API Routes
Endpoints for managing focus sessions, settings, and distraction blocking
"""
from datetime import datetime, timedelta
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel, Field, field_validator

from app.core.database import get_db
from app.core.rate_limiter import limiter, api_rate_limit, sensitive_rate_limit
from app.core.validators import (
    validate_time_string, validate_list_items, sanitize_string,
    check_xss, MAX_SHORT_TEXT
)
from app.api.routes.auth import get_current_user
from app.models.user import User
from app.services.focus_service import FocusService

router = APIRouter(tags=["Focus"])


# ═══════════════════════════════════════════════════════════════════
# REQUEST/RESPONSE MODELS
# ═══════════════════════════════════════════════════════════════════

class FocusSettingsUpdate(BaseModel):
    default_blocked_apps: Optional[List[str]] = Field(None, max_length=100)
    default_blocked_websites: Optional[List[str]] = Field(None, max_length=100)
    allowed_apps: Optional[List[str]] = Field(None, max_length=100)
    allowed_websites: Optional[List[str]] = Field(None, max_length=100)
    auto_start_from_calendar: Optional[bool] = None
    auto_detect_gaps: Optional[bool] = None
    min_gap_minutes: Optional[int] = Field(None, ge=5, le=120)
    focus_duration_minutes: Optional[int] = Field(None, ge=5, le=240)
    break_duration_minutes: Optional[int] = Field(None, ge=1, le=60)
    long_break_duration_minutes: Optional[int] = Field(None, ge=5, le=120)
    sessions_before_long_break: Optional[int] = Field(None, ge=1, le=10)
    break_reminders_enabled: Optional[bool] = None
    blocking_mode: Optional[str] = Field(None, pattern=r"^(soft|hard|strict)$")
    bypass_password: Optional[str] = Field(None, max_length=100)
    work_start_time: Optional[str] = Field(None, pattern=r"^([01]?[0-9]|2[0-3]):[0-5][0-9]$")
    work_end_time: Optional[str] = Field(None, pattern=r"^([01]?[0-9]|2[0-3]):[0-5][0-9]$")
    work_days: Optional[List[int]] = None

    @field_validator('default_blocked_apps', 'default_blocked_websites', 'allowed_apps', 'allowed_websites')
    @classmethod
    def validate_app_lists(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        if v is None:
            return v
        validated = []
        for item in v:
            item = sanitize_string(item, MAX_SHORT_TEXT)
            if len(item) > 200:
                raise ValueError(f"Item too long: {item[:20]}...")
            validated.append(item)
        return validated

    @field_validator('work_days')
    @classmethod
    def validate_work_days(cls, v: Optional[List[int]]) -> Optional[List[int]]:
        if v is None:
            return v
        for day in v:
            if day < 0 or day > 6:
                raise ValueError("Work days must be between 0 (Sunday) and 6 (Saturday)")
        return list(set(v))

    @field_validator('bypass_password')
    @classmethod
    def validate_bypass_password(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        is_safe, error = check_xss(v)
        if not is_safe:
            raise ValueError(error)
        return v


class CreateFocusBlockRequest(BaseModel):
    start_time: datetime
    end_time: datetime
    title: str = Field(default="Focus Time", max_length=100)
    blocking_enabled: bool = True
    sync_to_calendar: bool = False

    @field_validator('title')
    @classmethod
    def validate_title(cls, v: str) -> str:
        v = sanitize_string(v, MAX_SHORT_TEXT)
        is_safe, error = check_xss(v)
        if not is_safe:
            raise ValueError(error)
        return v


class StartFocusRequest(BaseModel):
    block_id: Optional[str] = None  # If None, starts ad-hoc session


class DistractionBlockedRequest(BaseModel):
    block_id: str = Field(..., max_length=100)
    app_or_site: str = Field(..., max_length=200)

    @field_validator('app_or_site')
    @classmethod
    def validate_app_or_site(cls, v: str) -> str:
        v = sanitize_string(v, 200)
        is_safe, error = check_xss(v)
        if not is_safe:
            raise ValueError(error)
        return v


class FocusSettingsResponse(BaseModel):
    id: str
    default_blocked_apps: List[str]
    default_blocked_websites: List[str]
    allowed_apps: List[str]
    allowed_websites: List[str]
    auto_start_from_calendar: bool
    auto_detect_gaps: bool
    min_gap_minutes: int
    focus_duration_minutes: int
    break_duration_minutes: int
    long_break_duration_minutes: int
    sessions_before_long_break: int
    break_reminders_enabled: bool
    blocking_mode: str
    work_start_time: str
    work_end_time: str
    work_days: List[int]
    total_focus_minutes: int
    total_distractions_blocked: int
    current_streak_days: int
    longest_streak_days: int

    class Config:
        from_attributes = True


class FocusBlockResponse(BaseModel):
    id: str
    start_time: datetime
    end_time: datetime
    duration_minutes: int
    title: str
    status: str
    blocking_enabled: bool
    blocked_apps: List[str]
    blocked_websites: List[str]
    completed_minutes: int
    success_rate: Optional[float]
    distractions_blocked: int

    class Config:
        from_attributes = True


class CalendarGapResponse(BaseModel):
    start_time: datetime
    end_time: datetime
    duration_minutes: int
    quality_score: int


class FocusSuggestionResponse(BaseModel):
    start_time: datetime
    end_time: datetime
    duration_minutes: int
    reason: str
    priority: str


# ═══════════════════════════════════════════════════════════════════
# SETTINGS ENDPOINTS
# ═══════════════════════════════════════════════════════════════════

@router.get("/settings", response_model=FocusSettingsResponse)
@limiter.limit(api_rate_limit())
async def get_focus_settings(
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get user's focus mode settings"""
    service = FocusService(db)
    settings = await service.get_or_create_settings(current_user.id)
    return settings


@router.put("/settings", response_model=FocusSettingsResponse)
@limiter.limit(sensitive_rate_limit())
async def update_focus_settings(
    request: Request,
    updates: FocusSettingsUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update user's focus mode settings"""
    service = FocusService(db)
    settings = await service.update_settings(
        current_user.id,
        updates.model_dump(exclude_unset=True)
    )
    return settings


# ═══════════════════════════════════════════════════════════════════
# FOCUS BLOCK ENDPOINTS
# ═══════════════════════════════════════════════════════════════════

@router.get("/blocks", response_model=List[FocusBlockResponse])
async def get_focus_blocks(
    hours_ahead: int = Query(default=24, ge=1, le=168),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get upcoming scheduled focus blocks"""
    service = FocusService(db)
    blocks = await service.get_upcoming_focus_blocks(current_user.id, hours_ahead)
    return blocks


@router.post("/blocks", response_model=FocusBlockResponse)
async def create_focus_block(
    request: CreateFocusBlockRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new focus block"""
    service = FocusService(db)

    if request.end_time <= request.start_time:
        raise HTTPException(status_code=400, detail="End time must be after start time")

    block = await service.create_focus_block(
        user_id=current_user.id,
        start_time=request.start_time,
        end_time=request.end_time,
        title=request.title,
        blocking_enabled=request.blocking_enabled,
        sync_to_calendar=request.sync_to_calendar
    )
    return block


@router.delete("/blocks/{block_id}")
async def delete_focus_block(
    block_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Delete a focus block"""
    from app.models import FocusBlock
    from sqlalchemy import select, and_

    result = await db.execute(
        select(FocusBlock).where(
            and_(
                FocusBlock.id == block_id,
                FocusBlock.user_id == current_user.id
            )
        )
    )
    block = result.scalar_one_or_none()

    if not block:
        raise HTTPException(status_code=404, detail="Focus block not found")

    await db.delete(block)
    await db.commit()

    return {"message": "Focus block deleted", "id": block_id}


# ═══════════════════════════════════════════════════════════════════
# ACTIVE SESSION ENDPOINTS
# ═══════════════════════════════════════════════════════════════════

@router.get("/active", response_model=Optional[FocusBlockResponse])
async def get_active_session(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get currently active focus session"""
    service = FocusService(db)
    block = await service.get_active_focus_session(current_user.id)
    return block


@router.post("/start", response_model=FocusBlockResponse)
@limiter.limit(sensitive_rate_limit())
async def start_focus_session(
    http_request: Request,
    request: StartFocusRequest = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Start a focus session"""
    service = FocusService(db)

    # Check if already in focus
    active = await service.get_active_focus_session(current_user.id)
    if active:
        raise HTTPException(status_code=400, detail="Already in an active focus session")

    try:
        block = await service.start_focus_session(
            current_user.id,
            block_id=request.block_id if request else None
        )
        return block
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/end/{block_id}", response_model=FocusBlockResponse)
async def end_focus_session(
    block_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """End a focus session"""
    service = FocusService(db)

    try:
        block = await service.end_focus_session(current_user.id, block_id)
        return block
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/pause/{block_id}", response_model=FocusBlockResponse)
async def pause_focus_session(
    block_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Pause a focus session"""
    service = FocusService(db)

    try:
        block = await service.pause_focus_session(current_user.id, block_id)
        return block
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


# ═══════════════════════════════════════════════════════════════════
# SUGGESTIONS & GAPS
# ═══════════════════════════════════════════════════════════════════

@router.get("/gaps", response_model=List[CalendarGapResponse])
async def get_calendar_gaps(
    days_ahead: int = Query(default=7, ge=1, le=30),
    min_gap_minutes: int = Query(default=30, ge=15, le=240),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Find gaps in calendar suitable for focus time"""
    service = FocusService(db)

    now = datetime.now()
    end_date = now + timedelta(days=days_ahead)

    gaps = await service.detect_calendar_gaps(
        current_user.id,
        now,
        end_date,
        min_gap_minutes
    )

    return [
        CalendarGapResponse(
            start_time=gap.start_time,
            end_time=gap.end_time,
            duration_minutes=gap.duration_minutes,
            quality_score=gap.quality_score
        )
        for gap in gaps
    ]


@router.get("/suggestions", response_model=List[FocusSuggestionResponse])
async def get_focus_suggestions(
    days_ahead: int = Query(default=7, ge=1, le=14),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get AI-generated suggestions for focus blocks"""
    service = FocusService(db)
    suggestions = await service.get_focus_suggestions(current_user.id, days_ahead)

    return [
        FocusSuggestionResponse(
            start_time=s.start_time,
            end_time=s.end_time,
            duration_minutes=s.duration_minutes,
            reason=s.reason,
            priority=s.priority
        )
        for s in suggestions
    ]


# ═══════════════════════════════════════════════════════════════════
# STATS & BLOCKING
# ═══════════════════════════════════════════════════════════════════

@router.get("/stats")
async def get_focus_stats(
    days: int = Query(default=7, ge=1, le=90),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get focus statistics"""
    service = FocusService(db)
    stats = await service.get_focus_stats(current_user.id, days)
    return stats


@router.post("/distraction-blocked")
async def record_distraction_blocked(
    request: DistractionBlockedRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Record a blocked distraction attempt"""
    service = FocusService(db)
    await service.record_distraction_blocked(
        current_user.id,
        request.block_id,
        request.app_or_site
    )
    return {"message": "Distraction recorded"}


@router.get("/check-auto-start")
async def check_auto_start(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Check if there's a focus event that should auto-start"""
    service = FocusService(db)
    result = await service.check_auto_start_focus(current_user.id)
    return {"should_start": result is not None, "event": result}


# ═══════════════════════════════════════════════════════════════════
# QUICK ACTIONS
# ═══════════════════════════════════════════════════════════════════

@router.post("/quick-start")
@limiter.limit(sensitive_rate_limit())
async def quick_start_focus(
    request: Request,
    duration_minutes: int = Query(default=50, ge=5, le=180),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Quick start a focus session with custom duration"""
    service = FocusService(db)

    # Check if already in focus
    active = await service.get_active_focus_session(current_user.id)
    if active:
        raise HTTPException(status_code=400, detail="Already in an active focus session")

    now = datetime.now()
    end_time = now + timedelta(minutes=duration_minutes)

    block = await service.create_focus_block(
        user_id=current_user.id,
        start_time=now,
        end_time=end_time,
        title="Quick Focus",
        blocking_enabled=True
    )

    # Immediately start it
    block = await service.start_focus_session(current_user.id, block.id)

    return block


@router.post("/schedule-from-suggestion")
async def schedule_from_suggestion(
    start_time: datetime,
    end_time: datetime,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Schedule a focus block from a suggestion"""
    service = FocusService(db)

    block = await service.create_focus_block(
        user_id=current_user.id,
        start_time=start_time,
        end_time=end_time,
        title="Scheduled Focus",
        blocking_enabled=True,
        sync_to_calendar=True
    )

    return block
